### DISCUSSION TITLE
GoPro 11 - Burn Text Overlay ISO & Shutter Speed

### DISCUSSION THREAD

**Discussion by stryker1080** [gopro/labs#695] [Q&A]
Hi,

After trying for about 2 hours and doing about 100 various permutations of the code, I have given up and need to ask on the forum LOL.
I want to be able to "burn in" text in the bottom of my videos which displays the current ISO and camera Shutter Speed. I was able to modify and tweak the examples provided, and this does technically work:

https://gopro.github.io/labs/control/set/?cmd=oMBRNO=10oMBURN=%22(0,100)[BC%27ISOE%27%20%27SHUT%27]%22D&title=Overlays

Couple of Questions:

In the examples on GitHub (https://[gopro.github.io/labs/control/overlays/](https://gopro.github.io/labs/control/overlays/), which has ISO listed, the ISOE GPMF data is followed by %d. I read the documentation and i figured this tells it how to format the display, with %d as 64-bit double. I don't want the ISO to be displayed with 3 decimal points, so i tried changing %d to %J for a 64bit unsigned integer, but it didn't display anything. Just a 'J' letter instead of numbers. Is there any way to change the formatting of the GPMF data?

For the Shutter Speed, I really would like it to display 1/100 instead of 0.01, which is what gets displayed with SHUT.
I tried multiple variables of 1/SHUT , or 1/'SHUT', trying to create a variable D and assign SHUT or 'SHUT' to it and take the inverse.... no luck. Is there any way to do some math and take the inverse/reciprocal of SHUT and display that? 

Finally, after each of the numbers for ISO and Shutter speed, it follows this with the actual META DATA Code. So in the Video, it looks like this:  400.00ISOE 0.016SHUT
Is there any way to make it so it doesn't automatically add the tag in the burn-in?

I would like the end result to look like this:
ISO=400 SHUT=1/120

Thank you for any help or guidance... I just felt like i was spinning in circles.

**Comment by dnewman-gpsw** [comment]
That nice formatting you seek is not currently possible. The Labs overlay engine is old and it is predates the new scripting that support math and variables, which could attempt to reformat the data.  

Why do you want to do this?  The metadata is available after capture, so you can get this information, without impacting image quality.

**Comment by stryker1080** [comment]
Good to know regarding the formatting.

I saw a previous post where you shared a bit of a hack on hack trick, and doing some math outside the MBURN tag and displaying the variable using a loop. I might try that tonight and see if it suits my needs.

The reason i would like to do this is to be able to film some test footage (for tutorials or figuring out ND filters) for various lighting conditions and riding conditions. Would like to know what the camera is doing during the video so have easy reference visual with current settings.

Do you know of any easier way to get the metadata after the fact and add it in as text in post production?

**Reply by dnewman-gpsw** [reply]
There are tools for developer, open source tools for extraction (see https://gopro.github.io/gpmf-parser/.)  You can also display them using commercial tools like https://goprotelemetryextractor.com/telemetry-overlay-gps-video-sensors

You can used oMEXPS=1 to render them to the LCD.  And on HERO12 you can script the capture of the LCD for a really crude solution: !1N!F!R  Once per second you get:
![GRAB-20240102-172345a](https://github.com/gopro/labs/assets/10856084/4bbf1bdd-7824-40e8-afbc-0cfb464c5c0a)

**Comment by stryker1080** [comment]
Thank you, i will give those a try. I did get something partially working, much closer anyway:

oMBRNO=10=Ii=Ss=S^-1oMBURN="(0,100)[BCISO:$I SHUTTER:1/$S]"!R9

![Go Pro ISO SHUTTER OVERLAY](https://github.com/gopro/labs/assets/6558824/1a4c8e6b-d096-429d-8764-6be21df0a042)

For some reason the modulus command was getting ISO to display weird values below 100. Saw in some other post that supposedly it would remove decimal points, which it did, but seems more accurate with the code above. Will post back if i get better solutions.

Thanks for all the help!

**Reply by dnewman-gpsw** [reply]
I've forgotten the hacks of hacks, thanks for the reminder.

This works: =A-1oMBRNO=10=Ii=I&A=Ss=S&AoMBURN="(0,100)[BC ISO:$I SHUT:1/$S]"!R15

While =S^-1 might have worked (in C/C++).  1) minus number imports aren't supported. 2) ^ is not an XOR, is powf().

The above works. A is initialized to 0, so '=A-1' is A--; setting A to -1.  T?hen '=I&A' is I&=-1 forcing the integer.  Same as =S&A.   

![image](https://github.com/gopro/labs/assets/10856084/7711a2a4-1bf7-4fd7-b85a-55e0c7da0a54)

**Comment by stryker1080** [comment]
Hi,

Thanks for the response back! Yes, that was a good post with the hack of hacks. 

For my understanding, could you explain how the code you have and the following (not sure if i understand):

=A-1                                     // A is initialized to zero and we subtract from it to get -1, make sense
Side Question, are the variables Signed/Unsigned Int, Floats, ?

=Li                                    // Stores ISO value in L
= L&A                             // What does this do? I understand this as: (float)((int)L & A)

Is the AND operator a bitwise operator, or is it using it like a logical operator? When i did some microcontroller programming, I would use bitwise AND OR operators to mask or toggle various bits to changes settings, etc; but struggle to understand what it does in this situation. 

If the ISO was say 125.78, i can see the INT typcast will be usefull, so it becomes 125. If logical, then 125 & -1 = FALSE. If Bitwise, then what is -1 in binary? Does it used signed bit infront or two complement? If two complement, then -1 would be 1111111111. Then L&A just becomes L, except in its INT typecase. Is that correct?

=Ss                        // This stores the shutter speed in S, makes sense

=S&A                 // This does not make sense to me. (float)((int)S & A)

In almost all cases, the shutter speed will be some small number such as 0.1 or 0.01, etc. If you take int(0.1), wont that result in 0? Sorry, just reading about converting decimals to binary.... If i understand, if we had S=0.125 
then 0.125 x 2 = 0 + 0.25
        0.25 x 2 =   0 + 0.5
        0.5  x 2  = 1
which becomes 10 00 = 8 + 0 + 0 + 0 = 8 which is exactly what i want! 

Is that how that works?

**Reply by dnewman-gpsw** [reply]
All variables are internally stored as floats, but are display to the precision needed.  So 1.0 will display at 1, 1.001 will display as 1.001 etc.

'=L&A' is L=(float)((int)L&A)

AND is bitwise operator.

As AND required a integer, the & or -1 is only to cast to int, as the bitwise 11111 AND does nothing else.  Hack within hack to truncate float point numbers.

From https://gopro.github.io/labs/control/actions/
''' 
s shutterValue - >sValueCMD - testing shutter, where 1/Value is used for shutter speed 
'''

The value of $s is 120 for a shutter speed of 1/120.  So it much easier to work with.

Display raw shutter speed:   **"SHUT: 1/$s"!R**
Display in milliseconds:  **=A-1=Ss=T1=T/S=T*1000"SHUT: $Tms"!R4**
Display integer shutter speed: **=A-1=Ss=S&A"SHUT: 1/$S"!R4**

It is weird language, compressed for QR codes.

**Comment by stryker1080** [comment]
That makes a lot of sense, thank you so much for explaining and answering my questions. Lots of really good nuggets here :) 

This opens up lots of doors and possibilities!!
